### ISSUE TITLE
"using private name" error in augmentation of `export =` namespace

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#14121]
**TypeScript Version:** nightly (2.3.0-dev.20170216)

**Code**

```ts
// @declaration: true

// @Filename: /node_modules/foo/index.d.ts
export = foo;
declare namespace foo {
    export type T = number;
}

// @Filename: /a.ts
import * as foo from "foo";
declare module "foo" {
    export function f(): T;
}
```

**Expected behavior:**

No error.

**Actual behavior:**

`/a.ts(3,26): error TS4060: Return type of exported function has or is using private name 'T'.`

**Comment by landonpoch** [comment]
Not sure if this provides any additional useful data but I'm seeing similar behavior when augmenting a module:

```
declare module "superagent" {
    export interface SuperAgentRequest {
        /**
         * This feature is part of the "superagent" lib however, it is missing 
         * from "@types/superagent" so we need to extend it.
         */
        retry(maxRetries?: number): SuperAgentRequest;
    }
}
```
`error TS4057: Return type of method from exported interface has or is using private name 'SuperAgentRequest'.`

V2.2.2

**Comment by aluanhaddad** [comment]
FYI, you should not export the augmented declaration. When you declare an augmentation like that, `SuperAgentRequest` is an `interface` or perhaps a `class` whose type you are modifying in place.

**Comment by landonpoch** [comment]
@aluanhaddad - Thanks for the tip.  However, I only added the export keyword after I got that error message in the first place.  In other words, either way it breaks.

**Comment by aluanhaddad** [comment]
Yeah, I know. If it is a valid augmentation, then it is validly re-exported but I remember reading somewhere on here that it should be an error and may become one in a future version so I thought I would give you a heads up.

**Comment by aluanhaddad** [comment]
@landonpoch Ah I see the problem, the declaration does not export it https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/superagent/index.d.ts#L15

In fact... it doesn't export anything. The entire declaration is bug-ridden.

**Comment by landonpoch** [comment]
Since it is an interface inside of the request module, which gets exported at the very end, shouldn't that be considered exported?  Kind of like how you were saying before that I shouldn't have included the export keyword on the augmentation I was doing?

**Comment by landonpoch** [comment]
@aluanhaddad Thanks for updating the typings file.  I pulled in the new file as you have modified it and I'm still seeing the same issue.  Again, this is occurring with the export keyword or without the export keyword in my module augmentation code above.  One more thing to note, this only happens when I have the declaration flag set to true.  The odd thing is though, SuperAgentRequest is only used internally to the file I'm writing.  I never even try and export it in any way.  Weird...

**Comment by aluanhaddad** [comment]
It needs to be exported in the sense of exported from the module which is the file in which it resides. As a declaration, it is useless otherwise.
Restart your IDE, etc. I was able to augment without errors.
Basically, modules and exports are tied to URLs in ECMAScript. So it needs to get exported for you to consume it.
No module can cause another to exports its contents, transitively.

**Comment by aluanhaddad** [comment]
@landonpoch so I made a silly mistake. I missed the obvious error (too used to looking at UMD typings)
You need to write it like this.
```ts
declare module 'superagent' {
  namespace request {
      interface SuperAgentRequest {
        /**
         * This feature is part of the "superagent" lib however, it is missing
         * from "@types/superagent" so we need to extend it.
         */
        retry(maxRetries?: number): SuperAgentRequest;
    }
  }
}
```
It is essentially declaration merging so it needs to be declared in an augmentation of the containing scope

**Comment by landonpoch** [comment]
@aluanhaddad when nesting inside the request namespace I get: [ts] Property 'retry' does not exist on type 'SuperAgentRequest.'

When I don't put it in the namespace then the IDE picks up the augmented method just fine.  I think it's because the request namespace is in fact the sole item exported from the module in the first place.  superagent is effectively an alias to the request export to begin with.  Also, restarting the IDE made no difference on my ability to transpile with the declaration flag set to true.

All very worthy things to try.  I am however still suspecting that this is related to this open bug.

**Comment by aluanhaddad** [comment]
Right, I was talking about the declaration that is installed from npm. Since it doesn't export the Types, drillling down was necessary, you are working with the `.d.ts` I edited, I am not. Try
```bash
$ npm i typescript@next
```
I get no errors and that is with `--declaration`

**Comment by landonpoch** [comment]
@aluanhaddad, Still seeing the issue after upgrading to @next.  Not sure how it could be working for you.  If I get time I might try and produce the smallest possible steps to reproduce.

**Comment by aluanhaddad** [comment]
Could be your moduleResolution, or baseUrl, or paths, or types, or typeRoots, or....
Just try installing it into an empty project.

**Comment by mhegazy** [comment]
I do not think that `T` should be accessible in the augmentation in the first place.

**Comment by weswigham** [comment]
@mhegazy There is no longer a visibility error as @andy-ms describes (not sure when it was fixed, but I'll add a regression test - https://github.com/Microsoft/TypeScript/pull/22748), and @DanielRosenwasser thinks the current behavior makes sense (certainly it aligns with how namespaces work).
